Services/CustomDeviceModels.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Exceptions;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.StorageAdapter;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services
+{
+    public interface ICustomDeviceModels
+    {
+        /// <summary>
+        /// Get list of custom device models.
+        /// </summary>
+        Task<IEnumerable<DeviceModel>> GetListAsync();
+        
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        Task<DeviceModel> GetAsync(string id);
+        
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        Task<DeviceModel> InsertAsync(DeviceModel deviceModel);
+
+        /// <summary>
+        /// Create or replace a custom device model.
+        /// </summary>
+        Task<DeviceModel> UpsertAsync(DeviceModel deviceModel);
+
+        /// <summary>
+        /// Delete a custom device model.
+        /// </summary>
+        Task DeleteAsync(string id);
+    }
+    public class CustomDeviceModels : ICustomDeviceModels
+    {
+        private const string STORAGE_COLLECTION = "deviceModels";
+
+        private readonly IStorageAdapterClient storage;
+        private readonly ILogger log;
+
+        public CustomDeviceModels(
+            IStorageAdapterClient storage,
+            ILogger logger)
+        {
+            this.storage = storage;
+            this.log = logger;
+        }
+
+        /// <summary>
+        /// Get list of custom device models.
+        /// </summary>
+        public async Task<IEnumerable<DeviceModel>> GetListAsync()
+        {
+            var data = await this.storage.GetAllAsync(STORAGE_COLLECTION);
+            var results = new List<DeviceModel>();
+            foreach (var item in data.Items)
+            {
+                var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+                deviceModel.ETag = item.ETag;
+                deviceModel.Type = "CustomModel";
+                results.Add(deviceModel);
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> GetAsync(string id)
+        {
+            var item = await this.storage.GetAsync(STORAGE_COLLECTION, id);
+            var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+            deviceModel.ETag = item.ETag;
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel)
```

Comment:
Suggest renaming this to UpdateAsync since that is what it is calling internally